src/NuGet.Protocol.Core.v2/DownloadResourceV2.cs

```diff
@@ -38,27 +39,29 @@ public DownloadResourceV2(V2Resource resource)
 
             return Task.Run(() =>
             {
-                var version = SemanticVersion.Parse(identity.Version.ToString());
+                token.ThrowIfCancellationRequested();
+
                 try
                 {
-                    var package = V2Client.FindPackage(identity.Id, version);
+                    var sourcePackage = identity as SourcePackageDependencyInfo;
+                    var dataServiceRepo = V2Client as DataServicePackageRepository;
 
-                    if (package != null)
+                    if (sourcePackage != null 
+                        && dataServiceRepo != null
+                        && sourcePackage.PackageHash != null
+                        && sourcePackage.DownloadUri != null)
                     {
-                        if (V2Client is UnzippedPackageRepository)
-                        {
-                            var packagePath = Path.Combine(V2Client.Source, identity.Id + "." + version);
-                            var directoryInfo = new DirectoryInfo(packagePath);
-                            if (directoryInfo.Exists)
-                            {
-                                return new DownloadResourceResult(package.GetStream(), new PackageFolderReader(directoryInfo));
-                            }
-                        }
-
-                        return new DownloadResourceResult(package.GetStream());
+                        // If this is a SourcePackageDependencyInfo object with everything populated 
+                        // and it is from an online source, use the machine cache and download it using the
+                        // given url.
+                        // If this info is not provided fallback to the old method.
```

Comment:
not clear from the comment where is the fallback. I assume it is the code after the else?